### DISCUSSION TITLE
Handling missing pages

### DISCUSSION THREAD

**Discussion by VladoPortos** [tarampampam/error-pages#140] [Q&A] [answered]
Is there a way to handle missing pages with some variable ?

The issue is, when the endpoint/service does not exist in Traefik it will correctly return 404 nice images because it will go for 'query: /{status}.html' so 404.html from container error-pages.

Eg when call http://page.not.setup.com

But when I call  http://page.not.setup.com/something  I get basically redirected to /something on `error-pages` image, and this return just a simple message from the in build go http server that that page is missing. Is there a way to automatically redirect inside the container everything that is missing to the themed 404.html ? 

Basically, I need this part of code: https://github.com/tarampampam/error-pages/blob/e2489a24879586bd087b2de0a091cd1723c7f00e/internal/http/handlers/notfound/handler.go

To redirect to 404.html, unfortunately I don't know `go` that well to mod it.

**Comment by tarampampam** [comment]
It looks like there is something wrong with the trarfik settings. Could you please attach your config?

**Comment by VladoPortos** [comment]
Sure. It might be actually issue with Traefik, I would assume that I need some kind of middleware that redirects not found pages to 404.html...

I have it setup like this:

Deployment
```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: error-pages
  namespace: traefik
  labels:
    app: error-pages
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
  selector:
    matchLabels:
      app: error-pages
  template:
    metadata:
      labels:
        app: error-pages
    spec:
      containers:
      - name: error-pages
        image: tarampampam/error-pages:2.18.0
        imagePullPolicy: IfNotPresent
        env:
        - name: TEMPLATE_NAME
          value: "app-down"
        - name: SHOW_DETAILS
          value: "true"
        ports:
        - containerPort: 8080
        securityContext:
          readOnlyRootFilesystem: true
          runAsNonRoot: true
          allowPrivilegeEscalation: false
          # Run the container as nobody:nogroup
          runAsUser: 65534
          runAsGroup: 65534
          capabilities:
            drop:
            - NET_RAW
        resources:
          requests:
            cpu: 50m
            memory: 16Mi
          limits:
            memory: 32M
```

Service:

```
---
apiVersion: v1
kind: Service
metadata:
  name: error-pages
  namespace: traefik
spec:
  selector:
    app: error-pages
  ports:
  - name: http
    port: 8080
    protocol: TCP
  type: ClusterIP
```

Ingress
```
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: error-pages-http-ingressroute
  namespace: traefik
spec:
  entryPoints:
    - web
  routes:
    - kind: Rule
      match: HostRegexp(`{host:.+}`)
      priority: 1
      services:
        - kind: Service
          name: error-pages
          port: 8080
```
Ingress is setup to match any hostname that was not matched by anything else, thus the low priority,

Middleware
```
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: error-pages-middleware
  namespace: traefik
spec:
  errors:
    status:
      - "400-599"
    query: /{status}.html
    service:
      name: error-pages
      port: 8080
```

Now when I have another ingress
```
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: argo-route  
  labels:
    app: traefik
  namespace: argocd
spec:
  entryPoints:
  - websecure
  routes:
  - match: Host(`some.domain.com`)
    kind: Rule
    services:
      - name: argo-cd-argocd-server
        port: 80
 ```

As of that it works fine. However when I edit the ingress for argocd and change the port to 81, which makes it not reachable. I get to 'some.domain.com' and get nice 404 from the error-pages container... however if I enter `some.domain.com/something' or 'some.domain.com/somehting.html' this get passed to the error-pages container and of course these endpoints does not exist there. Thus the message from my initial message. Maybe I'm doing it wrong ? 

I know that If I map the middleware to the ingress for argocd it self then it should work ok ( did not test yet), but what about described situation above... ? It could be solved if the error-pages http server it self refer to its own 404.html when page is not found on it...

**Comment by VladoPortos** [comment, marked as answer]
Hmm thinking about it now,

I should probably add:

```
    middlewares:
      - name: error-pages-middleware
```

To the `catch all` ingress route, so it can reference it self with error codes...

**Reply by VladoPortos** [reply]
This was exactly it, added the middleware to the `deafult` catch all ingress and all is working as expected :)

**Reply by tarampampam** [reply]
:+1:

**Reply by ScreNN** [reply]
I have the same situation. Can you help me?

Sometimes my domain is accessed with a subdomain.

`subdomain.domain.com`

But this page does not exist at my domain
In this case the browser returns a standard 404. But code-pages does not react to it. Is there any way to fix it?

Part of the config

Traefik.yml:
```yml
################################################################
#                        EntryPoints                           #
################################################################
entryPoints:
  http:
    address: :80
    http:
      middlewares:
        - error-pages@file
        - ip2location@file
      redirections:
        entryPoint:
          to: https
          scheme: https

  https:
    address: :443
    http:
      middlewares:
        - error-pages@file
        - ip2location@file
```

Router:
```yml
http:
  routers:
    error-pages:
      rule: "HostRegexp(`{host:.+}`)"
      entryPoints:
        - "https"
      service: error-pages
      priority: 10

  services:
    error-pages:
      loadBalancer:
        servers:
          - url: "http://IP:PORT"
```

Middlewares:
```yml
  # Error Pages
    error-pages:
      errors:
        status:
          - "400-599"
        service: error-pages
        query: /{status}.html
```
